test/scala/services/ValidatorPoolTest.scala

```diff
@@ -0,0 +1,157 @@
+package services
+
+import model.{Category, ResponseRule, RuleMatch}
+import org.scalatest._
+import org.scalatest.concurrent.PatienceConfiguration.{Timeout}
+import org.scalatest.concurrent.ScalaFutures
+import org.scalatest.time.SpanSugar._
+import utils.Validator
+import utils.Validator.ValidatorResponse
+
+import scala.concurrent.{ExecutionContext, Promise}
+
+/**
+  * A mock validator to test the pool implementation. Doesn't
+  * complete work until markAsComplete is called, to test
+  * queue behaviour.
+  */
+class MockValidator(id: Int) extends Validator {
+  private var currentWork: Option[Promise[ValidatorResponse]] = None
+  private var currentResponses:  Option[ValidatorResponse] = None
+  def getId = s"mock-validator-$id"
+  def getCategory = "mock-validator-cat"
+  def getRules = List.empty
+
+  def check(request: ValidatorRequest) = {
+    val promise = Promise[ValidatorResponse]()
+    val future = promise.future
+    currentWork = Some(promise)
+    maybeComplete()
+    future
+  }
+
+  def markAsComplete(responses: ValidatorResponse): Unit = {
```

Comment:
@jennygrahamjones this is something I'm a little confused by -- at the moment, `check` can be called _after_ `markAsComplete`, which before I added code in mitigation could cause e.g. the `'complete queued works'` spec to fail. But my Scala/Java is not good enough to understand why! I'd love to pair to understand why this might be happening.

Reply:
e.g. in

```
val checkFuture = pool.check("test-1", "Example text")
validators.foreach(_.markAsComplete(responses))
```

`check` calls a validator for each category in turn, but `markAsComplete` is called (I think!) before `check` is done calling each validator.